### ISSUE TITLE
Huge memory consumption when using get_dft_array() or output_dft() under MPI

### ISSUE THREAD

**Issue by FiodarM** [NanoComp/meep#1707]
I use meep with python interface under MPI on a cluster with 50-100 processes. With certain test parameters memory usage during simulation is ~25GB and `run()` finishes successfully. After the simulation I try to save the frequency domain fields to HDF5 using `output_dft()`. On `output_dft()` rapid increase in memory usage to >100GB occurs which exceeds the memory allocated for the simulation so I lose the simulation results. I also tried with `get_dft_array()` and saving to HDF5 manually with h5py and experienced the same problem. It is important that when using `get_dft_array()` memory leak occurs on `get_dft_array()` operation itself i.e, before the saving to HDF5. So the issue seems not to be related to HDF5. As I understand, `output_dft()` method just calls corresponding C++ method and creates no python objects. So this issue, as I can see, is neither related to python interface.

Is such behavior normal when using many MPI processes or is this a problem with meep?

I built meep from source from master branch.

Thanks in advance.

**Comment by FiodarM** [comment]
The recorded memory consumption vs time chart:
![ram-use](https://user-images.githubusercontent.com/11776788/127226602-d0c75b32-f9df-4c79-b06c-b333ac39731d.png)

**Comment by FiodarM** [comment]
The same problem happens with `sim.get_array(mp.Dielectric)` as well.

**Comment by smartalecH** [comment]
cc @oskooi 

I often see similar behavior (although never this extreme). Usually each "field dump" operation (e.g. when pulling fields after forward and adjoint runs) induces a spike in memory usage that's 25-35% the current consumption (e.g. from 150 GB to 200 GB and then back down).

This *might* be due to the gathering that happens when the user makes a call like this. Simply put, *every* proc receives a copy of *all* the DFT fields (this is certainly the case when using the python interface). The distributed-memory paradigm breaks down at these IO junctions.

We've talked about getting around this for adjoint optimization (where the forward and adjoint fields are always locally stored on each proc that owns them, and then the final recombination step is *also* performed locally). It might be nice to generalize that approach (e.g. when you just want to dump the fields and aren't necessarily doing adjoint optimization).

An even better solution (IMHO) is to use the hybrid multithreading/multiprocessing approach in #1628. This only requires one process per node.

**Comment by stevengj** [comment]
For `get_dft_array`, every process gets a copy of *all* the DFT data, so that will potentially be a huge amount of memory.

For `output_dft`, however, each process should ideally write only its portion of the data to the HDF5 file.

**Comment by stevengj** [comment]
It looks like the `output_dft` implementation (cc @HomerReid) currently does not use parallel I/O (https://github.com/NanoComp/meep/blob/ef2c9cad8a148c2602fa60e065a12352a89d6c21/src/dft.cpp#L1129) — that means that it gathers all of the data to the master process in order to write it.

This should really be fixed so that each process only computes and writes its own portion of the DFT data.

**Comment by FiodarM** [comment]
I also noted another strange behavior. When I call `get_epsilon()` at the beginning of the simulation (only once):
* the memory consumption increases and does not drop even after explicit `del` of the array and `gc.collect()`;
* the simulation speed drops as if it is single-core.
![ram-use](https://user-images.githubusercontent.com/11776788/127374527-36d57876-15f8-42c1-88f2-ead0ed7670a0.png)

Two last steps in green line correspond to calls of `get_dft_array()` for two field components.

**Comment by FiodarM** [comment]
@stevengj @HomerReid Is it crucial to set `parallel` to `false` for dft output? Will it work if one just changes it to `false` to `true` in this line?
https://github.com/NanoComp/meep/blob/ef2c9cad8a148c2602fa60e065a12352a89d6c21/src/dft.cpp#L1129

**Comment by stevengj** [comment]
No, simply changing `false` to `true` is not sufficient — one would have to also change the subsequent logic to ensure that each process only writes its "own" data to the file.
